hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -0,0 +1,458 @@
+          doHoodieMetadataTableValidationOnce();
+          break;
+        default:
+          LOG.info("Unsupported running mode [" + cfg.runningMode + "], quit the job directly");
+      }
+    } catch (Exception e) {
+      throw new HoodieException("Unable to do hoodie metadata table validation in " + cfg.basePath, e);
+    } finally {
+
+      if (asyncMetadataTableValidateService.isPresent()) {
+        asyncMetadataTableValidateService.get().shutdown(true);
+      }
+    }
+  }
+
+  private void doHoodieMetadataTableValidationOnce() {
+    doMetaTableValidation();
+  }
+
+  private void doHoodieMetadataTableValidationContinuous() {
+    asyncMetadataTableValidateService.ifPresent(service -> {
+      service.start(null);
+      try {
+        service.waitForShutdown();
+      } catch (Exception e) {
+        throw new HoodieException(e.getMessage(), e);
+      }
+    });
+  }
+
+  public void doMetaTableValidation() {
+    metaClient.reloadActiveTimeline();
+    String basePath = metaClient.getBasePath();
+    HoodieSparkEngineContext engineContext = new HoodieSparkEngineContext(jsc);
+
+    List<String> allPartitions = validatePartitions(engineContext, basePath);
+
+    for (String partitionPath : allPartitions) {
+      validateFilesInPartition(engineContext, partitionPath);
+    }
+
+    LOG.info("MetaTable Validation Success.");
+  }
+
+  /**
+   * Compare the listing partitions result between metadata table and fileSystem.
+   */
+  private List<String> validatePartitions(HoodieSparkEngineContext engineContext, String basePath) {
+    // compare partitions
+    List<String> allPartitionPathsFromFS = FSUtils.getAllPartitionPaths(engineContext, basePath, false, cfg.assumeDatePartitioning);
+    List<String> allPartitionPathsMeta = FSUtils.getAllPartitionPaths(engineContext, basePath, true, cfg.assumeDatePartitioning);
+
+    if (!compareCollectionResult(allPartitionPathsFromFS, allPartitionPathsMeta)) {
+      String message = "Compare Partitions Failed! " + "AllPartitionPathsFromFS : " + allPartitionPathsFromFS + " and allPartitionPathsMeta : " + allPartitionPathsMeta;
+      LOG.error(message);
+      throw new HoodieValidationException(message);
+    }
+
+    return allPartitionPathsMeta;
+  }
+
+  /**
+   * Compare the listing files result between metadata table and fileSystem.
+   * For now, validate two kinds of apis:
+   * 1. getLatestFileSlices
+   * 2. getLatestBaseFiles
+   * @param engineContext
+   * @param partitionPath
+   */
+  private void validateFilesInPartition(HoodieSparkEngineContext engineContext, String partitionPath) {
+    HoodieTableFileSystemView metaFsView = null;
+    HoodieTableFileSystemView fsView = null;
+
+    try {
+      metaFsView = createHoodieTableFileSystemView(engineContext, true);
+      fsView = createHoodieTableFileSystemView(engineContext, false);
+
+      validateLatestFileSlices(metaFsView, fsView, partitionPath);
+      validateLatestBaseFiles(metaFsView, fsView, partitionPath);
+
```

Comment:
Let's also validate `getAllFileGroups()`, i.e., all file groups, and all file slices within file groups, besides the latest, to make sure that older file slices can still be listed correctly with metadata table.

Reply:
+1 good call out. but validation will be intensive. but we might need it. but if someone is always interested only in snapshot queries, latest base file validation should be good right? 
thinking if this can be made configurable too.

Reply:
Yes, this should be configurable.

Reply:
Okay, changed.
Add getAllFileGroups() validation and add three new options 
1. --validate-latest-fileSlices
2. --validate-latest-baseFiles
3. --validate-all-file-groups

to control which kinds of validation to use.

Reply:
Sg.  As a side note, `MergeOnReadIncrementalRelation` uses `fsView.getAllFileGroups(partitionPath)` to get all file groups on the read path, so the validation of all file groups based on metadata table is still critical.

Reply:
Added `fsView.getAllFileGroups(partitionPath)` check. PTAL :)